por/porserver.go

```diff
@@ -59,104 +110,111 @@ func GetPorServer() *PorServer {
-		return err
+func (ps *PorServer) GetOrComputeVrf(data []byte) ([]byte, []byte, error) {
+	if v, ok := ps.vrfCache.Get(data); ok {
+		if res, ok := v.(*vrfResult); ok {
+			return res.vrf, res.proof, nil
+		}
 	}
 
-	nxPk, err := sc.GetLastPubkey()
+	vrf, proof, err := ed25519.GenerateVrf(ps.account.PrivKey(), data)
 	if err != nil {
-		log.Error("Get last public key error:", err)
-		return err
+		return nil, nil, err
+	}
+
+	res := &vrfResult{
+		vrf:   vrf,
+		proof: proof,
 	}
 
-	if !common.IsEqualBytes(dcPk, nxPk) {
-		return errors.New("it's not the right signer")
+	ps.vrfCache.Add(data, res)
+
+	return vrf, proof, nil
+}
+
+func (ps *PorServer) Sign(relayMessage *pb.Relay, nextPubkey, prevNodeID []byte, mining bool) error {
+	vrf, _, err := ps.GetOrComputeVrf(relayMessage.BlockHash)
+	if err != nil {
+		log.Error("Get or compute VRF error:", err)
+		return err
 	}
 
-	err = sc.Sign(ps.id, nextPubkey, mining, ps.account.PubKey(), ps.account.PrivKey())
+	signature, err := pb.ComputeSignature(vrf, relayMessage.LastSignature, ps.id, nextPubkey, mining)
 	if err != nil {
-		log.Error("Signature chain signing error:", err)
+		log.Error("Computing signature error:", err)
 		return err
 	}
 
+	ps.sigChainElemCache.Add(signature, &sigChainElemInfo{
+		nextPubkey:    nextPubkey,
+		prevNodeID:    prevNodeID,
+		prevSignature: relayMessage.LastSignature,
+		mining:        mining,
+		blockHash:     relayMessage.BlockHash,
+	})
+
+	relayMessage.LastSignature = signature
+	relayMessage.SigChainLen++
+
 	return nil
 }
 
-func (ps *PorServer) Verify(sc *SigChain) error {
+func (ps *PorServer) Verify(sc *pb.SigChain) error {
 	if err := sc.Verify(); err != nil {
 		return err
 	}
 
 	return nil
 }
 
-func (ps *PorServer) CreateSigChain(dataSize uint32, dataHash, blockHash *common.Uint256, srcID,
-	destPubkey, nextPubkey []byte, mining bool) (*SigChain, error) {
-	return NewSigChain(ps.account.PubKey(), ps.account.PrivKey(), dataSize, dataHash[:], blockHash[:], srcID, destPubkey, nextPubkey, mining)
-}
-
-func (ps *PorServer) CreateSigChainForClient(dataSize uint32, dataHash, blockHash *common.Uint256, srcID,
-	srcPubkey, destPubkey, signature, proof []byte, sigAlgo SigAlgo) (*SigChain, error) {
+func (ps *PorServer) CreateSigChainForClient(dataSize uint32, blockHash *common.Uint256, srcID,
+	srcPubkey, destPubkey, signature []byte, sigAlgo pb.SigAlgo) (*pb.SigChain, error) {
 	pubKey, err := ps.account.PubKey().EncodePoint(true)
 	if err != nil {
 		log.Error("Get account public key error:", err)
 		return nil, err
 	}
-	sigChain, err := NewSigChainWithSignature(dataSize, dataHash[:], blockHash[:],
-		srcID, srcPubkey, destPubkey, pubKey, signature, proof, sigAlgo, false)
+	sigChain, err := pb.NewSigChainWithSignature(dataSize, blockHash[:], srcID, srcPubkey, destPubkey, pubKey, signature, sigAlgo, false)
 	if err != nil {
 		log.Error("New signature chain with signature error:", err)
 		return nil, err
 	}
+	ps.srcSigChainCache.Add(signature, sigChain)
```

Comment:
It seems that this `signature` NOT signed by Client, it just an util.RandomBytes(32) from [ws.localNode.SendRelayMessage](https://github.com/nknorg/nkn/pull/502/commits/a5564bc5d7d790eb13be20d1d6344707080a4194#diff-b9add2abd88d41a8ecdb9a966dd818daL39)
Please correct me if I have incorrect understanding.

Reply:
You are right. Client side has not implemented signature yet :)

Reply:
Maybe we should left a `// TODO` comments at here.

Reply:
Don't worry, it will be done in the next few weeks (if not days) 🤣